components/ai_chat/core/browser/conversation_handler.cc

```diff
@@ -0,0 +1,1385 @@
+  // When panel is open or has existing conversation, do not change the state.
+  // if (!is_conversation_active_ && chat_history_.empty()) {
+  if (chat_history_.empty()) {
+    SetShouldSendPageContents(false);
+  }
+}
+
+void ConversationHandler::AddSubmitSelectedTextError(
+    const std::string& selected_text,
+    mojom::ActionType action_type,
+    mojom::APIError error) {
+  if (error == mojom::APIError::None) {
+    return;
+  }
+  const std::string& question = GetActionTypeQuestion(action_type);
+  mojom::ConversationTurnPtr turn = mojom::ConversationTurn::New(
+      CharacterType::HUMAN, action_type,
+      mojom::ConversationTurnVisibility::VISIBLE, question, selected_text,
+      std::nullopt, base::Time::Now(), std::nullopt, false);
+  AddToConversationHistory(std::move(turn));
+  SetAPIError(error);
+}
+
+void ConversationHandler::OnFaviconImageDataChanged() {
+  for (const mojo::Remote<mojom::ConversationUI>& client :
+       conversation_ui_handlers_) {
+    client->OnFaviconImageDataChanged();
+  }
+}
+
+void ConversationHandler::OnUserOptedIn() {
+  MaybePopPendingRequests();
+  MaybeFetchOrClearContentStagedConversation();
+}
+
+void ConversationHandler::AddToConversationHistory(
+    mojom::ConversationTurnPtr turn) {
+  if (!turn) {
+    return;
+  }
+
+  chat_history_.push_back(std::move(turn));
+
+  OnHistoryUpdate();
+}
+
+void ConversationHandler::PerformAssistantGeneration(
+    const std::string& input,
+    std::string page_content /* = "" */,
+    bool is_video /* = false */,
+    std::string invalidation_token /* = "" */) {
+  auto data_received_callback =
+      base::BindRepeating(&ConversationHandler::OnEngineCompletionDataReceived,
+                          weak_ptr_factory_.GetWeakPtr());
+
+  auto data_completed_callback =
+      base::BindOnce(&ConversationHandler::OnEngineCompletionComplete,
+                     weak_ptr_factory_.GetWeakPtr());
+
+  bool should_refine_page_content =
+      features::IsPageContentRefineEnabled() &&
+      page_content.length() > GetMaxContentLengthForModel(GetCurrentModel()) &&
+      input != l10n_util::GetStringUTF8(IDS_AI_CHAT_QUESTION_SUMMARIZE_PAGE);
+  if (should_refine_page_content && associated_content_delegate_) {
+    DVLOG(2) << "Asking to refine content, which is of length: "
+             << page_content.length();
+    associated_content_delegate_->GetTopSimilarityWithPromptTilContextLimit(
+        input, page_content, GetMaxContentLengthForModel(GetCurrentModel()),
+        base::BindOnce(&ConversationHandler::OnGetRefinedPageContent,
+                       weak_ptr_factory_.GetWeakPtr(), input,
+                       std::move(data_received_callback),
+                       std::move(data_completed_callback), page_content,
+                       is_video));
+    return;
+  } else if (!should_refine_page_content && is_content_refined_) {
+    is_content_refined_ = false;
+    OnAssociatedContentInfoChanged();
+  }
```

Comment:
If we shouldn't refine content, but we did refine content on the previous submission, then this property of SiteInfo has changed, so we notify the client via `OnAssociatedContentInfoChanged`.
It can happen if the page content got smaller or the model was changed.
I added a comment to make it clearer.